KAFKA-7286: Avoid getting stuck loading large metadata records

flavray opened the pull request:
If a group metadata record size is higher than
offsets.load.buffer.size, loading offsets and group metadata from
__consumer_offsets would hang forever.

This was due to a buffer being too small to fit any message bigger than
the maximum configuration. If this happens, temporarily use a buffer
that can fit the large records and move on.

Commit:
KAFKA-7286: Avoid getting stuck loading large metadata records

If a group metadata record size is higher than
offsets.load.buffer.size, loading offsets and group metadata from
__consumer_offsets would hang forever.

This was due to a buffer being too small to fit any message bigger than
the maximum configuration. If this happens, temporarily use a buffer
that can fit the large records and move on.

vvcephei commented on core/src/main/scala/kafka/coordinator/group/GroupMetadataManager.scala:
```diff
@@ -524,9 +524,15 @@ class GroupMetadataManager(brokerId: Int,
-              MemoryRecords.readableRecords(buffer)
+              val recordsBuffer = if (fileRecords.sizeInBytes > buffer.capacity) {
+                debug(s"Loaded offsets and group metadata from $topicPartition with buffer larger than configured offsets.load.buffer.size")
+                ByteBuffer.allocate(fileRecords.sizeInBytes)
```
Is it ok to allocate a new buffer larger than the configured size for this case? What if we instead logged an error message and threw an exception?

flavray replied:
Just logging an error and throwing an exception seems fine to me as well. Thought it was fine to temporarily allocate more memory if needed to read a message (even if that makes the `offsets.load.buffer.size` parameter less true, and some systems might run Kafka with tight memory limits) as the above `log.read` has `minOneMessage = true`.
No strong opinion on any side, as long as loading the partition is not silently stuck 🙂

vvcephei replied:
Thanks for the explanation. I'm not in a position to have an opinion either!

@ijuma, can someone else review this?

lambdaliu commented on core/src/main/scala/kafka/coordinator/group/GroupMetadataManager.scala:
```diff
@@ -524,9 +524,15 @@ class GroupMetadataManager(brokerId: Int,
-              fileRecords.readInto(buffer, 0)
-              MemoryRecords.readableRecords(buffer)
+              val recordsBuffer = if (fileRecords.sizeInBytes > buffer.capacity) {
+                debug(s"Loaded offsets and group metadata from $topicPartition with buffer larger than configured offsets.load.buffer.size")
```
It would be better if we change the log level to WARN and add the value of fileRecords.sizeInBytes to the log message.

dhruvilshah3 commented:
Thanks for the patch, @flavray. The patch looks reasonable to me.

I think another way to implement this fix could be to enforce the condition that we do not exceed the configured maximum metadata size when appending to the log in `GroupMetadataManager#storeGroup`. It looks like we already have some validation in `GroupMetadataManager#storeOffsets` so may be we are missing this in `storeGroup`? This would mean that users will see this error when metadata exceeds the configured limit though.

cc @hachikuji as he should have more context about this.

Commit:
Raise logging level to warn

Also add number of bytes allocated in the log message

flavray commented:
@dhruvilshah3 it looks like the `offset.metadata.max.bytes` is 4096 by default. As far as I can tell, group metadata messages can get a lot larger than this.

I don't have that much context on these configs though, so I'm definitely fine waiting for a reviewer with more experience on the subject (@hachikuji as you mentioned?) to move forward with this approach!

flavray commented:
Hi there, just want to follow-up to see if any progress can be made here 🙂 
Cheers!

dhruvilshah3 commented on core/src/main/scala/kafka/coordinator/group/GroupMetadataManager.scala:
```diff
@@ -524,9 +524,16 @@ class GroupMetadataManager(brokerId: Int,
-              buffer.clear()
-              fileRecords.readInto(buffer, 0)
-              MemoryRecords.readableRecords(buffer)
+              val sizeInBytes = fileRecords.sizeInBytes
```
Can we move the buffer allocation here so that we don't have to reallocate when we need more space?

flavray replied:
I can definitely see the benefits of moving allocation here, however doing this would require to change from `lazy val buffer` to `var buffer` in [L513](https://github.com/apache/kafka/blob/4a97eafda55c369763fba9a7b3e81aa2f59dd59f/core/src/main/scala/kafka/coordinator/group/GroupMetadataManager.scala#L513).

This means allocating `buffer` even if it it not needed (in case `fetchDataInfo.records` are `MemoryRecords`, the buffer is not allocated).
That should not be a big deal to have this buffer pre-allocated every time and grow it if needed, just wanted to point out the "downsides" of the approach.
Do you have any opinion @dhruvilshah3 ? 🙂

dhruvilshah3 replied:
Hmm, it seems like you should be able to do something like this:

```
            case fileRecords: FileRecords =>
              val sizeInBytes = fileRecords.sizeInBytes
              val recordsBuffer = ByteBuffer.allocate(Math.max(config.loadBufferSize, sizeInBytes))
              if (recordsBuffer.capacity > config.loadBufferSize)
                warn("...")
              fileRecords.readInto(recordsBuffer, 0)
              MemoryRecords.readableRecords(recordsBuffer)
```

flavray replied:
Oh I see, smart! Wouldn’t this be allocating a new buffer at every loop though?

dhruvilshah3 replied:
Ahh, I didn't notice that. Makes sense to keep it the way you have it then.

dhruvilshah3 reviewed (approved):
Thanks for the PR. Left a minor comment, LGTM otherwise.

hachikuji commented on core/src/main/scala/kafka/coordinator/group/GroupMetadataManager.scala:
```diff
@@ -524,9 +524,16 @@ class GroupMetadataManager(brokerId: Int,
+              val sizeInBytes = fileRecords.sizeInBytes
+              val recordsBuffer = if (sizeInBytes > buffer.capacity) {
+                warn(s"Loaded offsets and group metadata from $topicPartition with buffer larger ($sizeInBytes bytes) than configured offsets.load.buffer.size")
+                ByteBuffer.allocate(sizeInBytes)
```
I'm wondering if we should just use the oversized buffer for the remainder of the loading since we have already allocated it. For one thing, that would allow the other buffer to be collected immediately. It may also prevent further allocations if the topic has many oversized messages.

vvcephei replied:
Does this mean that we will basically ignore the `offsets.load.buffer.size`, or, rather, that it becomes the "initial" size instead of the limit?

In that case, should we update the docs on the config?

hachikuji replied:
The config becomes a soft limit in any case with this patch. I think it does make sense to mention this in the documentation.

flavray replied:
I've updated the branch with a mean that replaces the buffer in case we need to allocate a bigger one. This was also suggested by @dhruvilshah3 it may be a better idea anyway.
The code does not look as idiomatic as it used to, but at least it "does the job" 🙂I am definitely open to make changes if either of you has a better suggestion!

hachikuji commented on core/src/test/scala/unit/kafka/coordinator/group/GroupMetadataManagerTest.scala:
```diff
@@ -1606,7 +1638,7 @@ class GroupMetadataManagerTest {
     val offsetCommitRecords = createCommittedOffsetRecords(committedOffsets, apiVersion = apiVersion, retentionTime = Some(100))
     val memberId = "98098230493"
-    val groupMetadataRecord = buildStableGroupRecordWithMember(generation, protocolType, protocol, memberId, apiVersion)
+    val groupMetadataRecord = buildStableGroupRecordWithMember(generation, protocolType, protocol, memberId, 0, apiVersion)
```
nit: I think we could depend on the default value for `assignmentSize` here. We just need to name the `apiVersion` argument explicitly.

flavray replied:
Good catch! I've fixed this 🙂

hachikuji reviewed (commented):
Thanks for the patch. Good find. Left a couple comments.

Commit:
Depend on the default assignmentSize in test

Commit:
Improve buffer allocation when loading large metadata records

Instead of temporary using a larger buffer, use this buffer for later
records as well.

hachikuji commented on core/src/main/scala/kafka/coordinator/group/GroupMetadataManager.scala:
```diff
@@ -524,7 +526,18 @@ class GroupMetadataManager(brokerId: Int,
+
+              if (buffer.capacity < bytesNeeded) {
+                if (config.loadBufferSize < bytesNeeded)
+                  warn(s"Loaded offsets and group metadata from $topicPartition with buffer larger ($bytesNeeded bytes) than configured offsets.load.buffer.size")
```
nit: may as well add the configured value to the message?

hachikuji commented on core/src/main/scala/kafka/coordinator/group/GroupMetadataManager.scala:
```diff
@@ -524,7 +526,18 @@ class GroupMetadataManager(brokerId: Int,
             case fileRecords: FileRecords =>
-              buffer.clear()
+              val sizeInBytes = fileRecords.sizeInBytes
+              val bytesNeeded = Math.max(config.loadBufferSize, sizeInBytes)
```
A comment may be helpful to mention that because of the `minOneMessage` behavior, we may need to grow the buffer to ensure progress can be made.

flavray replied:
Good point, I've added a comment. Let me know if you think it needs to be improved

hachikuji reviewed (commented):
@flavray Thanks for the updates. I had a couple small additional comments. I think it would also be good to update the documentation as you suggested before.

And one final thing: we have similar loading logic in `TransactionStateManager.loadTransactionMetadata`. The transaction state messages tend to be much smaller than the group metadata messages, so it's probably not a big deal, but I thought we may as well make the logic consistent there. We can also open a JIRA and do this separately if you prefer.

Commit:
Document the new behaviour of offsets.load.buffer.size

Commit:
Do not get stuck loading large transaction metadata records

flavray commented:
@hachikuji Thanks for the review!

I have replicated the behaviour for `TransactionMetdata` records as well. It adds a bit of code duplication, but I think that at this point (5-6 lines of code duplicated in two places), refactor would not help that much.

hachikuji reviewed (approved):
LGTM. Thanks for the patch!
